protocol-units/settlement/mcr/manager/src/manager.rs

```
@@ -0,0 +1,182 @@
+		let mut batch_acc = Vec::new();
+		loop {
+			tokio::select! {
+				Some(block_commitment) = receiver.recv(), if !ahead_of_settlement => {
+					commitments_to_settle.insert(
+						block_commitment.height,
+						block_commitment.commitment.clone(),
+					);
+					if block_commitment.height > max_height {
+						ahead_of_settlement = true;
+						let batch = mem::replace(&mut batch_acc, Vec::new());
+						if let Err(e) = client.post_block_commitment_batch(batch).await {
+							yield Err(e);
+							break;
+						}
+					}
+					batch_acc.push(block_commitment);
+				}
+				Some(res) = settlement_stream.next() => {
+					let settled_commitment = match res {
+						Ok(commitment) => commitment,
+						Err(e) => {
+							yield Err(e);
+							break;
+						}
+					};
+
+					let height = settled_commitment.height;
+					if let Some(commitment) = commitments_to_settle.remove(&height) {
+						let event = if commitment == settled_commitment.commitment {
+							BlockCommitmentEvent::Accepted(settled_commitment)
+						} else {
+							BlockCommitmentEvent::Rejected {
+								height,
+								reason: BlockCommitmentRejectionReason::InvalidCommitment,
+							}
+						};
+						yield Ok(event);
+					} else if let Some((&lh, _)) = commitments_to_settle.last_key_value() {
+						if lh < height {
+							// Settlement has left some commitments behind, but the client could
+							// deliver them of order?
+							todo!("Handle falling behind on settlement")
+						}
+					}
+					// Remove back-pressure if we can proceed settling new blocks.
+					if ahead_of_settlement {
+						let new_max_height = match client.get_max_tolerable_block_height().await {
+							Ok(h) => h,
+							Err(e) => {
+								yield Err(e);
+								break;
+							}
+						};
+						if new_max_height > max_height {
+							max_height = new_max_height;
+							ahead_of_settlement = false;
+						}
+					}
+				}
+				else => break
+			}
+		}
+	})
+}
```

If this works, for now it can get merged in as I don't want to slow this PR from being merged too much. But I think it could be refactors to not always `match` on results, instead update the Return types of the outer scope so that we can just bubble up the error with ?`. Then also, generaly reduce the nested ifs where reasonable

---

I'd like to use `?`, but it was not possible due to https://github.com/tokio-rs/async-stream/issues/63
There is a comment about this on top of the macro invocation.